File: client/src/edu/stanford/atom/sti/client/comm/io/TNetworkFileReader.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package edu.stanford.atom.sti.client.comm.io;

import edu.stanford.atom.sti.corba.Types.TNetworkFile;
import edu.stanford.atom.sti.corba.Types.TOctetSeqHolder;
/**
 *
 * @author Jason
 */
public class TNetworkFileReader {

    private TNetworkFile networkFile = null;

    private int length = 0;
    private boolean accessible = false;
    private byte[] data = null;
    private String filename = null;

    public TNetworkFileReader(TNetworkFile file) {
        networkFile = file;

        try {
            filename = networkFile.fileName();
            accessible = true;
        } catch (Exception e) {
            e.printStackTrace(System.out);
            accessible = false;
        }
    }

    public boolean read() {

        try {
            length = networkFile.length();
            filename = networkFile.fileName();
            accessible = true;
        } catch (Exception e) {
            e.printStackTrace(System.out);
            accessible = false;
        }

        if(!accessible || length <= 0)
            return false;

	data = new byte[length];

        boolean success = true;
        int i = 0;
        int packetSize = 1000;
        int timeout = 3;

        TOctetSeqHolder tOctet = new TOctetSeqHolder();

        while (i < length && timeout > 0) {
            try {
                networkFile.getBytes(i, packetSize, tOctet);
                success = true;
                accessible = true;
            } catch (Exception e) {
                e.printStackTrace(System.out);
                success = false;
                accessible = false;
                timeout--;
            }

            if (success && tOctet != null) {

                for (int j = 0; j < tOctet.value.length; j++) {
                    data[i + j] = tOctet.value[j];
                }
                i += packetSize;
            }
        }

        return success;
    }

    public byte[] getFileData() {
        return data;
    }

    public boolean isAccessible() {
        return accessible;
    }

    public int fileLength() {
        return length;
    }

    public String fileName() {
        return filename;
    }

}

File: device/src/TNetworkFileReader.cpp

#include <TNetworkFileReader.h>

TNetworkFileReader::TNetworkFileReader(STI::Types::TNetworkFile_ptr file)
{
	data = NULL;
	networkFileRef = file;
	
	try {
		fileName = networkFileRef->fileName();
		accessible = true;
	} catch(CORBA::TRANSIENT& ex) {
		accessible = false;
	} catch(CORBA::SystemException& ex) {
		accessible = false;
	} catch(CORBA::Exception&) {
		accessible = false;
	} catch(...) {
		accessible = false;
	}
}
TNetworkFileReader::~TNetworkFileReader()
{
	if(data != NULL ) {
		delete data;
		data = NULL;
	}
}

bool TNetworkFileReader::isAccessible()
{
	return accessible;
}

int TNetworkFileReader::fileLength()
{
	return length;
}

bool TNetworkFileReader::read()
{
	try {
		length = networkFileRef->length();
		accessible = true;
	} catch(CORBA::TRANSIENT& ex) {
		accessible = false;
	} catch(CORBA::SystemException& ex) {
		accessible = false;
	} catch(CORBA::Exception&) {
		accessible = false;
	} catch(...) {
		accessible = false;
	}
	if(!accessible || length <= 0)
		return false;

	if(data != NULL ) {
		delete data;
		data = NULL;
	}

	data = new char[length];
	
	STI::Types::TOctetSeq_var tOctetData;

	bool success = true;
	int i = 0;
	int packetSize = 1000;
	int timeout = 3;

	while(i < length && timeout > 0) {
		try {
			networkFileRef->getBytes(i, packetSize, tOctetData.out());
			success = true;
		} catch(...) {
			success = false;
			timeout--;
		}
		if(success) {

			for(unsigned j = 0; j < tOctetData->length(); j++) 
			{
				data[i + j] = tOctetData[j];
			}

			i += packetSize;
		}
	}

	return success;

}
